src/Database/Redis/Cluster.hs

```diff
@@ -263,17 +263,30 @@ evaluatePipeline refreshShardmapAction conn@(Connection shardNodeVar infoMap _)
         -- otherwise if there is any response in the connection buffer it'll get forwarded to other requests that are reusing the same connection.
         -- leading to jumbled up responses
         resps <- concat <$>
-          mapM (\(resp, (cc, r)) -> case resp of
-              Right v -> return v
-              Left (err :: SomeException) ->
-                case fromException err of
-                  Just (er :: TimeoutException) -> refreshShardmapAction >> throwIO er
-                  _ -> getRandomConnection cc conn >>= (`executeRequests` r)
+          mapM (\(resp, (nc, r)) -> do
+                responses <-  case resp of
+                                Right v -> return v
+                                Left (err :: SomeException) ->
+                                    case fromException err of
+                                        Just (er :: TimeoutException) -> hasLocked $ refreshShardmapAction Nothing >> throwIO er
+                                        _ -> getRandomConnection nc conn >>= (`executeRequests` r)
+                refreshedShardMapAndNodeConnsIORef <- IOR.newIORef Nothing
+                concat <$> mapM (\completedRequest@(CompletedRequest index request response) -> 
+                    if moved response then do
+                        map (CompletedRequest index request) <$> refreshShardMapAndRetryRequest refreshedShardMapAndNodeConnsIORef (hasLocked $ refreshShardmapAction (Just nc)) request
+                    else
+                        case askingRedirection response of
+                            Just (host,port) -> do 
+                                refreshedShardMapAndNodeConns <- fromMaybeM (hasLocked $ refreshShardmapAction (Just nc)) $ IOR.readIORef refreshedShardMapAndNodeConnsIORef
+                                maybeAskNode <- nodeConnWithHostAndPort refreshedShardMapAndNodeConns host port
+                                case maybeAskNode of
+                                    Just askNode -> map (CompletedRequest index request) <$> tail <$> requestNode askNode (["ASKING"] : [request])
+                                    Nothing -> do
+                                        map (CompletedRequest index request) <$> refreshShardMapAndRetryRequest refreshedShardMapAndNodeConnsIORef (hasLocked $ refreshShardmapAction (Just nc)) request
+                            Nothing -> return [completedRequest]
```

Comment:
So, apart from MOVED or AskingRedirection, we will be just giving the original response back to the Application ? Are there other possible errors that Redis might throw (came across TRYAGAIN, CROSSSLOT, CLUSTERDOWN, NOMASTERLINK) ?

Should this control block be rather : 
if moved
else if askingRedirection response = Just (host, port)
else (taking care of all other types of error replies by Redis) -> return [completedRequest] ?